openforcefield/typing/engines/smirnoff/forcefield.py

```diff
@@ -575,80 +631,175 @@ def parse(self, sources, input_format=None):
             # Make iterable object
             sources = [sources]
 
-        # Process all SMIRNOFF definition files or objects
-        # QUESTION: Allow users to specify forcefield URLs so they can pull forcefield definitions from the web too?
-        #trees = list()
-        if input_format is None:
-            # If we don't know, try all known formats
-            io_formats_to_try = self._parameter_io_handler_classes.keys()
-        else:
-            io_formats_to_try = [input_format]
-
+        # TODO: If a non-first source fails here, the forcefield might be partially modified
         for source in sources:
-            # TODO: Add support for other sources (other serialized formats, URLs, etc)
-            # Parse content depending on type
-            parse_successful = False
-
-            for parameter_io_format in io_formats_to_try:
-                parameter_io_handler = self.get_io_handler(parameter_io_format)
-                if type(source) == bytes:
-                    parameter_io_handler.parse_string(source)
-                    parse_successful = True
-                elif type(source) == str:
-                    #try: #TODO: What if it's the text of a XML file?
-                    #    byte_source = str.encode(source)
-                    #    parameter_io_handler.parse_string(byte_source)
-                    #    parse_successful = True
-                    #except:
-                    parameter_io_handler.parse_file(source)
-                    parse_successful = True
-                else:
-                    parameter_io_handler.parse_file(source)
-                    parse_successful = True
-                    # TODO: The try/except stuff here is dangerous. The io handler modifies the forcefield as it reads
-                    # the file. If it crashes midway through, the forcefield will be partially modified.
-                #except ParseError:
-                #    pass
-            if not (parse_successful):
-                valid_formats = [
-                    iohandler.format
-                    for iohandler in self._parameter_io_handlers
-                ]
-                msg = "Source {} does not appear to be in a known SMIRNOFF encoding.\n".format(
-                    source)
-                msg += "Valid formats are: {}".format(valid_formats)
-                raise Exception(msg)
-
-    # TODO : Move to ParameterIOHandler
-    def _parse_aromaticity_model(self, root):
-        """Parse the aromaticity model, make sure it is supported, and make sure it does not contradict previously-specified aromaticity models.
+            smirnoff_data = self.parse_smirnoff_from_source(source)
+            self.load_smirnoff_data(smirnoff_data)
+
+
+
+    def to_smirnoff_data(self):
```

Comment:
As a very minor consistency thing, I think in some places you refer to the output of this as smirnoff_data, and others as smirnoff_dict. Perhaps naming this method `to_dict` would be more consistent with the rest of your new code?

Reply:
I agree that this is bad, but won't change this right now. There are different ways we could standardize this (`__get/setstate__`, `from_dict`, `__init__`, designing for Pydantic in the future, etc) and it's big decision. Currently, there are conflicting serialization schemes throughout the toolkit (for example compare this to the `Serializable` class inheritance in `Molecule`). I think that the decision on serialization styles should be handled in a separate PR in the future.